File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& arr) {
        int currProduct=1,maxi=INT_MIN;
        for(int i=0;i<arr.size();i++){
            currProduct=currProduct*arr[i];
            maxi=max(currProduct,maxi);
            if(currProduct==0){
                currProduct=1;
            }
             
        }
        currProduct=1;
        for(int i=arr.size()-1;i>=0;i--){
            currProduct*=arr[i];
            maxi=max(maxi,currProduct);
            if(currProduct==0){
                currProduct=1;
            }
        }
        return maxi;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        curr=nums[0]
        curr_min=nums[0]
        curr_maxi=nums[0]
        result=nums[0]
        for i in range(1,len(nums)):
            temp_min=min(nums[i],curr_min*nums[i],curr_maxi*nums[i])
            temp_max=max(nums[i],curr_min*nums[i],curr_maxi*nums[i])
            curr_min=temp_min
            curr_maxi=temp_max
            result=max(curr_maxi,result)
        return result
